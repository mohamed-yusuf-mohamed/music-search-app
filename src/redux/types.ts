export interface ResponseData {
  artistName: string,
  kind: string,
  trackName: string, 
  artworkUrl60: string   
}

// TODO: types
export interface SearchBarState {
  loading: boolean,
  error: boolean,
  // status: 'loading' | 'loaded' | 'error',
  parsedInput: string,
  input: string,
  data: ResponseData[],
  fetchCount: number
}

const initialState: SearchBarState = {
  // status: ""
  loading: false,
  error: false,
  parsedInput: "",
  input: "",
  data: [],
  fetchCount: 1
};