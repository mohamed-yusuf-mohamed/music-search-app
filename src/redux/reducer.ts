// import { combineReducers } from 'redux'
import {ResponseData} from "./types"

// COUNTER REDUCER
// const counterReducer = (state = 0, { type }) => {
//   switch (type) {
//     case types.INCREMENT:
//       return state + 1
//     case types.DECREMENT:
//       return state - 1
//     case types.RESET:
//       return 0
//     default:
//       return state
//   }
// }

// INITIAL TIMER STATE
const initialState = {
    loading: false,
    parsedInput: "",
    input: "",
    data: [] as ResponseData[],
    fetchCount: 0,
    error: false
}

// TIMER REDUCER
// TODO: types
// TODO: refactor
const reducer = (state = initialState, action: any) => {
  // ...state.data
  // ...action.payload.data
  // action.payload.data
  
  switch (action.type) {

    
    case "FETCH_STARTED":
      return {
        ...state,
        loading: true
        // fetchCount: state.fetchCount + 1
      }
    case "TEST":
      return {
        ...state,
        fetchCount: state.fetchCount + 1,
        data: action.payload.data
        // fetchCount: state.fetchCount + 1
      }
    case "FETCH_SUCCESS":
      return {
        ...state,
        data: action.payload.data,
        // fetched: action.payload.data,
        loading: false,
        fetchCount: state.fetchCount + 1
      }
    case "FETCH_ERROR":
      return {
        ...state,
        error: true,
        loading: false,
        fetchCount: state.fetchCount - 1
      }
    case "CHANGE_INPUT":
      return {
        ...state,
        input: action.payload,
        parsedInput: action.payload.split(' ').join('+')
      }
    default:
      return state
  }
}

// COMBINED REDUCERS
// const reducers = {
//   counter: counterReducer,
//   timer: timerReducer,
// }

export default reducer

// export default combineReducers(reducers)
