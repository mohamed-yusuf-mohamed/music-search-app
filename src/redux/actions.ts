import Api from "../api"
import {Dispatch, State, GetState, Thunk} from "./store"
import { Action } from '@reduxjs/toolkit';
import { AnyAction } from '@reduxjs/toolkit';
import { ThunkDispatch } from "./store";

// TODO: change commit author and name on all commits

const api = new Api()

export const reset = () => ({
  type: "RESET"
})

// TODO: resets on a new input

export const load = () => ({
  type: "LOAD"
})

export const newSearch = () => (dispatch: Dispatch) => {
  // console.log('dispatch', dispatch);
  dispatch(reset())
  return (dispatch as ThunkDispatch)(fetchData())
}

export const fetchData = (): Thunk => async (dispatch: Dispatch, getState: GetState) => {
    const state = getState()
    const {parsedInput, page, data} = state
    const offset = page * 10
    // TODO: validate the offset is working as it should
    const url = `https://itunes.apple.com/search?term=${parsedInput}&offset=${offset}&limit=10`

    dispatch(load())

    try {
      const res = await api.fetch(url)
      return dispatch({
        type: "FETCH",
        payload: {
          data: {
            ...data,
            ...res
          }
        }
      })
    } catch(err) {
      // add your logger here for the error
      return dispatch({
        type: "ERROR",
        payload: {
          error: true,
          loading: false
        }
      })
    }
}

export const handleInput = (e: React.ChangeEvent<HTMLInputElement>): AnyAction => ({ 
  type: "INPUT", 
  payload: e.target.value 
})