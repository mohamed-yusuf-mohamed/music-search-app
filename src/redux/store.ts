import { configureStore, ThunkAction, Action, applyMiddleware } from '@reduxjs/toolkit';
import { combineReducers } from 'redux'
// import counterReducer from '../../DELETEME/counter/counterSlice';
import reducer from "./reducer"
import thunk from "redux-thunk"
import { createStore } from 'redux';

// const reducer = combineReducers({counterReducer, appReducer})
// import rootReducer from './reducers/index';

// Note: this API requires redux@>=3.1.0
// export const store = createStore(reducer, applyMiddleware(thunk));

export const store = createStore(reducer, applyMiddleware(thunk))

// export const store = configureStore({
//   reducer: {
//     counter: counterReducer,
//     app: appReducer
//   }
// });

// export const 

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
