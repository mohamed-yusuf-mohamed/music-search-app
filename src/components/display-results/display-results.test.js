import React from 'react';
import {render} from '@testing-library/react';
import SearchInput from "./index"
import {fetchData} from "redux/actions";
import { Provider } from 'react-redux';
import { store } from 'redux/store';
import mock from "./__mocks__/data.json"
import fetch from "jest-fetch-mock"
import { enableFetchMocks } from 'jest-fetch-mock'
enableFetchMocks()

describe("display results ", () => {
  beforeAll(async () => {
    // preload store with "justin" search data
    fetch.mockResponseOnce(() => Promise.resolve(JSON.stringify(mock)))
    await store.dispatch(fetchData())
  })

  const {getByTestId} = render(
    <Provider store={store}>
        <SearchInput />
    </Provider>
  );

  const component = getByTestId("display-results-component")

  it('renders', async () => {
    expect(component).toMatchSnapshot()
  });
})
