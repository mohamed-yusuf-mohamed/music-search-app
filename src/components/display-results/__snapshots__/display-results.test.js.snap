// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display results  renders 1`] = `
<div
  class="sc-bdnxRM kmKBKg"
  data-testid="display-results-component"
>
  <div
    class="sc-gtsrHT jJmHrj"
    id="display-results-content"
  >
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Don't Stop Believin'
          </div>
          <div
            class="sc-fujyAs"
          >
            song : Journey
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Austin Powers: The Spy Who Shagged Me
          </div>
          <div
            class="sc-fujyAs"
          >
            feature-movie : Jay Roach
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Love Sex Magic (feat. Justin Timberlake)
          </div>
          <div
            class="sc-fujyAs"
          >
            song : Ciara
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Winner (feat. Justin Timberlake & T.I.)
          </div>
          <div
            class="sc-fujyAs"
          >
            song : Jamie Foxx
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Winner (feat. Justin Timberlake & T.I.)
          </div>
          <div
            class="sc-fujyAs"
          >
            song : Jamie Foxx
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Austin Powers In Goldmember
          </div>
          <div
            class="sc-fujyAs"
          >
            feature-movie : Jay Roach
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Justin
          </div>
          <div
            class="sc-fujyAs"
          >
            song : Korn
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Holy Grail (feat. Justin Timberlake)
          </div>
          <div
            class="sc-fujyAs"
          >
            song : JAY-Z
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Carry Out (feat. Justin Timberlake)
          </div>
          <div
            class="sc-fujyAs"
          >
            song : Timbaland
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-dlnjwi cztOVI"
    >
      <div
        class="sc-hKFxyN jepVTX"
      >
        <div
          class="sc-eCApnc hipvbo"
        >
          ARTWORK
        </div>
      </div>
      <div
        class="sc-jSFjdj gyA-DLT"
      >
        <div
          class="sc-gKAaRy wCVaH"
        >
          <div
            class="sc-iCoGMd"
          >
            Despacito (feat. Justin Bieber)
          </div>
          <div
            class="sc-fujyAs"
          >
            song : Luis Fonsi & Daddy Yankee
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-pNWdM hSpYqw"
    >
      Scroll to see more
    </div>
  </div>
</div>
`;
